File: pyNeuralNetwork/neural_network/components/neuron.py
from typing import Tuple


class Neuron:
    """Class to represent a single `Neuron` in a neural network
    """

    def __init__(self, layer_id: int, row_id: int):
        """Constructor method

        Parameters
        ----------
        layer_id : int
            The layer of the network
        row_id : int
            The row in the layer
        """
        self._id = layer_id, row_id
        self._bias = 0.0
        self._value = None
        self.bias_gradients = []

    def set_bias(self, bias: float):
        """Setter method

        Parameters
        ----------
        bias : float
            The new bias
        """
        self._bias = bias

    def get_bias(self) -> float:
        """Getter method

        Returns
        -------
        float
            The bias
        """
        return self._bias

    def set_value(self, value: float):
        """Setter method

        Parameters
        ----------
        value : float
            The new value
        """
        self._value = value

    def get_value(self) -> float:
        """Getter method

        Returns
        -------
        float
            The value
        """
        return self._value

    def get_id(self) -> Tuple[int, ...]:
        """Getter method

        Returns
        -------
        Tuple[int, ...]
            The id
        """
        return self._id

    def __str__(self):
        return f"Neuron {self._id}"

File: javaNeuralNetwork/src/main/java/mvg/neural_network/components/Neuron.java
package mvg.neural_network.components;

import java.util.ArrayList;
import java.util.List;

/** Class to represent a single Neuron in a neural network.
 * @version 1.0.0
 * @since 1.0.0
 */
class Neuron {

    private final List<Integer> id = new ArrayList<>();
    private double bias = 0.0;
    private double value = Double.NaN;
    private final List<Double> biasGradients = new ArrayList<>();

    /** Constructor method.
     * @param layerId The id of the layer of the network.
     * @param rowId The row in the layer.
     */
    Neuron(int layerId, int rowId) {
        this.id.addAll(List.of(layerId, rowId));
    }

    /** Adds an element to the biasGradients list.
     *
     * @param biasGradient The bias gradient to be added.
     */
    void addBiasGradient(double biasGradient) {
        biasGradients.add(biasGradient);
    }

    /** Clears the biasGradients list.
     *
     */
    void clearBiasGradients() {
        biasGradients.clear();
    }

    /** Getter method for id.
     *
     * @return The id.
     */
    List<Integer> getId() {
        return id;
    }

    /** Getter method for bias.
     *
     * @return The bias.
     */
    double getBias() {
        return bias;
    }

    /** Setter method for bias.
     *
     * @param bias The new bias.
     */
    void setBias(double bias) {
        this.bias = bias;
    }

    /** Getter method for value.
     *
     * @return The value.
     */
    double getValue() {
        return value;
    }

    /** Setter method for value.
     *
     * @param value The new value.
     */
    void setValue(double value) {
        this.value = value;
    }

    /** Getter method for biasGradients.
     *
     * @return A copy of the biasGradients list.
     */
    List<Double> getBiasGradients() {
        return List.copyOf(biasGradients);
    }

    /** String method.
     *
     * @return String representation of the Neuron.
     */
    @Override
    public String toString() {
        return "Neuron %s".formatted(id.toString());
    }
}
